### ISSUE TITLE
How to use .on() and .dropped() when using a group.

### ISSUE THREAD

**Issue by dcworldwide** [SGrondin/bottleneck#54]
I had global event listeners before adopting a group limiter.

```
limiter.on('dropped', (dropped) => {
    console.warn("Dropped IC request", dropped)
})

debug && limiter.on('debug', function (message, data) {
    console.log(message, data)
})
``

Now with a group limiter, how do i use this event listeners?

If i create a listener per limiter, who do I clean them up when the linter is garbage collected from the group? (or are linters not gcd?)

**Comment by SGrondin** [comment]
```js
const group = new Bottleneck.Group( ... )

const doStuff = (id) => {
  const limiter = group.key(id)

  limiter.removeAllListeners('dropped')
  limiter.on('dropped', (dropped) => {
    console.warn("Dropped IC request", dropped)
  })

  if (debug) {
    limiter.removeAllListeners('debug')
    limiter.on('debug', function (message, data) {
      console.log(message, data)
    })
  }
}

doStuff( ... )
```

The Group garbage collects limiters that haven't been used in over 5 minutes. It will automatically collect all the event listeners too.

**Comment by SGrondin** [comment]
> Just to clarify my understanding, what is the importance of using limiter.removeAllListeners() before registering a new event handler?

Because those event handlers might have already been registered on that limiter. In `v2.2.0` I plan on adding a better way to do this kind of thing.

>I'm curious how removeAllListeners will work in a highly concurrent environment where a limiter within the group is used by multiple requests at the same time.

`removeAllListeners`, as well as triggering events within Bottleneck is 100% synchronous so it won't have surprising side effects.

>Also, just wondering if the 5 minute gc timeout is configurable. I think it's a sane value, but in my case we will have potentially 1000 group keys so I'd like to make sure we are aggressive with gc.

Yes it's configurable. You can pass the `timeout` option when creating a Group or call `group.updateSettings({ timeout: 60000 })` to update an existing group. Look for the `updateSettings()` section under `Group` in the Readme for more information.

1000 limiters is not a problem at all, limiters are *tiny* and you should be able to make tens or hundreds of thousands without a problem. You're typically going to run out of CPU executing all those tasks long before you run out of memory.

**Comment by dcworldwide** [comment]
Thank you. I ran some stress tests today trying to break thus library anf wass pleasantly surprised by it's durability and predictability. Great job. We will be going to production with it.

**Comment by SGrondin** [comment]
That's wonderful to hear! I'll be releasing `v2.2.0` very soon, it contains improvements that came out of your questions in this Issue. I'll let you know when it's available on npm!

**Comment by dcworldwide** [comment]
Thanks @SGrondin that's great news for us :)

**Comment by SGrondin** [comment]
I've just released `2.2.0`. You can now do:
```js
const group = new Bottleneck.Group( ... )

group.on('created', (limiter, key) => {
  // Preparing a limiter for use.
  // The 'created' event is triggered only once per limiter

  limiter.on('dropped', (dropped) => {
    console.warn("Dropped IC request", dropped)
  })

  if (debug) {
    limiter.on('debug', function (message, data) {
      console.log(message, data)
    })
  }
})
```
Enjoy!

**Comment by dcworldwide** [comment]
Great addition. That is nice and logical. Thanks for the great work
